feat: add cleanup handlers to prevent memory leaks on socket close

vinikjkkj opened the pull request:
Added `registerSocketEndHandler` to register cleanup callbacks that run when socket closes. Implemented cleanup in:
- Signal repository: clear migrated session cache and LID mapping
- Chats socket: clear sync timeout and placeholder resend cache
- Messages recv: clear retry/call offer/placeholder caches and identity assert debounce
- Messages send: clear user devices/peer sessions caches and media connection
- Event buffer: add destroy

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Refactor**
  * Unified end-of-connection lifecycle with registered end handlers to run coordinated cleanup and avoid memory leaks.
  * Added explicit close/destroy/clear operations for repositories, mappings, caches, event buffers, and retry managers to ensure reliable resource release and state reset.

* **New Features**
  * Exposed an internal resend cache for external access and added an API to register socket-end cleanup callbacks.

* **Types**
  * Cache/store interfaces optionally support a close/destroy method.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

vinikjkkj committed:
feat: add cleanup handlers to prevent memory leaks on socket close

Added `registerSocketEndHandler` to register cleanup callbacks that run when socket closes. Implemented cleanup in:
- Signal repository: clear migrated session cache and LID mapping
- Chats socket: clear sync timeout and placeholder resend cache
- Messages recv: clear retry/call offer/placeholder caches and identity assert debounce
- Messages send: clear user devices/peer sessions caches and media connection
- Event buffer: add destroy

purpshell reviewed (commented):
@jlucaso1 let's do a destructive benchmark test on bartender by returning a server close error in the middle of a benchmark

purpshell commented:
Merge conflicts and we'll merge!

purpshell commented on src/Socket/chats.ts:
```diff
@@ -1222,6 +1219,20 @@ export const makeChatsSocket = (config: SocketConfig) => {
 		}
 	})
 
+	registerSocketEndHandler(() => {
+		if (awaitingSyncTimeout) {
+			clearTimeout(awaitingSyncTimeout)
+			awaitingSyncTimeout = undefined
+		}
+
+		if (!config.placeholderResendCache && placeholderResendCache.close) {
+			placeholderResendCache.close()
+		}
+
+		syncState = SyncState.Connecting
+		privacySettings = undefined
+	})
```
@vinikjkkj

purpshell committed:
Merge branch 'master' into end-handlers

Resolves conflicts in:
- src/Socket/socket.ts: keep both registerSocketEndHandler (end-handlers)
  and fetchAccountReachoutTimelock/fetchNewChatMessageCap (master)
- src/Socket/messages-recv.ts: keep both registerSocketEndHandler usage
  with cleanup callback and pruneExpiredTcTokens / new sock destructured
  fields (issuePrivacyTokens, fetchAccountReachoutTimelock)

purpshell committed:
fix: tighten socket-end cleanup ergonomics

- messages-send: type mediaConn as Promise<MediaConnInfo> | undefined and
  drop the unsafe 'undefined as any' cast in the cleanup
- socket: await socket-end handlers and widen the handler type to
  void | Promise<void> so async cleanup rejections are caught and logged
- chats/messages-recv: centralize placeholderResendCache ownership in chats.ts
  (it was being created and closed in both modules, splitting state); expose
  it on the returned sock and have messages-recv reuse it
- event-buffer: clear flushPendingTimeout in destroy() to prevent a flush
  call landing on a torn-down buffer

purpshell committed:
Merge branch 'master' into end-handlers

Resolves conflict in src/Utils/message-retry-manager.ts: keep both the
clear() cleanup method (this branch) and saveBaseKey/hasSameBaseKey/
deleteBaseKey from master, and extend clear() to also drop baseKeys so
the new cache is released on socket close like the other state.
